homeassistant/helpers/device_registry.py

```diff
@@ -939,25 +1016,56 @@ def async_update_device(  # noqa: C901
 
             if add_config_entry_id not in old.config_entries:
                 config_entries = old.config_entries | {add_config_entry_id}
+                config_subentries = old.config_subentries | {
+                    add_config_entry_id: {add_config_subentry_id}
```

Comment:
Why should there be a None item in the subentries set? Why not have an empty set?

Reply:
I don't find this specification in the architecture discussion post. My assumption is that a device is always part of the main config entry, regardless of subentries, until the device is removed from the config entry.

Reply:
With the implementation in the PR, a device is removed when its config subentry is removed. With what you're proposed, a device will instead be adopted by the parent config entry when the subentry is removed.

Reply:
I would expect a device to be removed when the subentry that it belongs to is removed.

Reply:
> With the implementation in the PR, a device is removed when its config subentry is removed. With what you're proposed, a device will instead be adopted by the parent config entry when the subentry is removed.

I'm not sure that that is a consequence. We know the state before and after so we can decide to remove the device if the last subentry is removed, if we want that behavior.

Reply:
> We know the state before and after so we can decide to remove the device if the last subentry is removed, if we want that behavior.

OK. Please explain how we can make the decision without knowing if a device belongs to the main entry or not.

Reply:
If there's a single subentry id before the change and the change removes that subentry id, we remove the device.